packages/core/src/lib/relay/index.ts

```diff
@@ -96,40 +98,63 @@ class Relay extends GossipSub implements IRelay {
       return { recipients: [] };
     }
 
-    return this.publish(this.pubSubTopic, msg);
+    return this.gossipSub.publish(this.pubSubTopic, msg);
   }
 
   /**
    * Add an observer and associated Decoder to process incoming messages on a given content topic.
    *
    * @returns Function to delete the observer
    */
-  addObserver<T extends IDecodedMessage>(
-    decoder: IDecoder<T>,
+  public subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
     callback: Callback<T>
   ): () => void {
-    const observer = {
-      decoder,
-      callback,
-    };
-    const contentTopic = decoder.contentTopic;
+    const contentTopicToObservers = toObservers(decoders, callback);
 
-    pushOrInitMapSet(this.observers, contentTopic, observer);
+    for (const contentTopic of contentTopicToObservers.keys()) {
+      const currObservers = this.observers.get(contentTopic) || new Set();
+      const newObservers =
+        contentTopicToObservers.get(contentTopic) || new Set();
+
+      this.observers.set(contentTopic, union(currObservers, newObservers));
+    }
 
     return () => {
-      const observers = this.observers.get(contentTopic);
-      if (observers) {
-        observers.delete(observer);
+      for (const contentTopic of contentTopicToObservers.keys()) {
+        const currentObservers = this.observers.get(contentTopic) || new Set();
+        const observersToRemove =
+          contentTopicToObservers.get(contentTopic) || new Set();
+
+        const nextObservers = leftMinusJoin(
+          currentObservers,
+          observersToRemove
+        );
+
+        if (nextObservers.size) {
+          this.observers.set(contentTopic, nextObservers);
+        } else {
+          this.observers.delete(contentTopic);
+        }
       }
     };
   }
 
+  public unsubscribe(pubSubTopic: TopicStr): void {
+    this.gossipSub.unsubscribe(pubSubTopic);
+    this.gossipSub.topicValidators.delete(pubSubTopic);
+  }
```

Comment:
I'd remove this for now as the behaviour is likely to be unpredictable.

We will re-introduce a proper API as we implement pubsub sharding and ability to support subscribing to several pubsub topic at a given time.

Or is this part of the `PubSub` interface?

Reply:
Seems not so removing it.